### PR TITLE
Rework property filter to support infinite scroll

### LINKED ISSUE #4802: Autocomplete event & person properties
## Bug description

We have users who need to load megabytes of JSON to show the filter dropdown. As it is now, this blocks the entire insights interface until the data has loaded.

## Expected behavior

The interface loads instantly and we search/autocomplete properties on the server (and cache just the most common ones).

## Additional context

Person properties need special handling: https://github.com/PostHog/posthog/issues/4797

### PR DESCRIPTION
**Pull request by samwinslow**
## Changes

Rework property filter to support infinite scroll and the tabbed interface described in https://github.com/PostHog/posthog/issues/4535#issuecomment-865363879. This means boatloads of event property definitions are no longer loaded at once, helping performance.

**Feature flag:** `4267-taxonomic-property-filter`

Video demo:

Screen Cast 2021-07-09 at 9 23 25 AM

The "taxonomic" name is because I'd like to _eventually_ abstract this UI to show taxonomy tags on events/actions.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] Frontend/CSS is usable at 320px (iPhone SE) and decent at 360px (most phones)
- [ ] Breaking changes are backwards-compatible. Ensure old/new frontend requests work with new/old backends, and vice versa.

### CODE REVIEW

**Comment by samwinslow** [comment]
TODO: Make CSS usable for small screens (hover tooltip position is wonky). This may/may not be a blocking issue, though.

**Review by mariusandra — COMMENTED** [review]
My mouse has a fast scroll wheel. When I let it spin, this is what happens:



Probably not the desired effect :).

Other than that, it looks cool!

My only true feedback is to refactor the data fetching part into Kea. I know this sounds rather vague, so here are some pointers:
- Each TaxonomicPropertyFilter should get a `logic` that's connected to it. You'll probably need to pass it a custom `key`, which could even be autogenerated for each property filter (and then saved with the react component, probably in an useRef or useState.
- Then move all other `use*` calls into that logic by creating the required actions, reducers, etc.
- Do the same for `infiniteList`... and into the same logic.
- Use various combinations of `selectors` in this logic to fetch data from the different models, apply filters (search), and then just get one value as a result for the data to display. Feed that to the lists.

This is all a bit vague I know, but I'd be curious to see where you end up if you start with some implementation.

However if we have a nice clean separation of data and view (React), in the end, this filter will be much stabler than now.

**Comment by mariusandra** [comment]
Regarding data flows and why "data first" makes sense.

I just found a bug where I just couldn't find a property even though it was in the list:



(`$initial_referrer_url` vs `Initial Referrer URL`)

The solution is to use `filterOption` in the antd `<Select>`, but [that doesn't work](https://github.com/ant-design/ant-design/issues/27956) if you're using select option groups (`filterOption` is just ran twice, once for each group). Plus even if it would, the label would be a react component, which we'd need to parse data out of somehow.

If we go "data first" and instead have in a `logic` all the definitions or a way to query them, and the searchfield's value, we can do all this filtering before any data reaches the visible frontend... instead of relying on the hacks we have to rely on now.

This is just one of the many antd quirks that we're constantly working around (everything to do with forms is another one), and swapping out these hacks for a pure data-first approach IMO makes a lot of sense.

**Comment by samwinslow** [comment]
@mariusandra The last commit here is almost ready to go IMO — I just have a few questions about loaders.

- `loadMoreRows` on `InfiniteLoader` needs a Promise which returns results. After I dispatch the action, how do I wait around in React for results?
- Are breakpoints available in loaders, or just listeners? I tried adding a breakpoint directly in my loader (taking it from param after `{ props, values, actions }` — but it doesn't seem to exist there)

File: frontend/src/lib/components/PropertyFilters/components/InfiniteSelectResults.tsx [hunk 1]
```diff
@@ -0,0 +1,214 @@
+import React, { useEffect } from 'react'
```
Comment (mariusandra) [inline frontend/src/lib/components/PropertyFilters/components/InfiniteSelectResults.tsx:1]: This file could probably be broken up.

File: frontend/src/scenes/insights/BreakdownFilter/BreakdownFilter.js [hunk 2]
```diff
@@ -110,6 +110,7 @@ function Content({ breakdown, breakdown_type, onChange }) {
         >
             <TabPane tab="Property" key="property">
                 <PropertyFilter
+                    pageKey={`${breakdown}-${window.location.pathname}`}
```
Comment (mariusandra) [inline frontend/src/scenes/insights/BreakdownFilter/BreakdownFilter.js:113]: What did this addition fix?

**Review by mariusandra — CHANGES_REQUESTED** [review]
I have some random comments inline, but for the questions and bigger comments:

- Yep, breakpoints are in loader, as the second argument of the function, just like for listeners. You're taking it from the wrong place.
- Let's avoid `InfiniteLoader` actually and refactor it to a standard `List`:

1. Just use a [regular `<List />`](https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md). 
2. Use its `onRowsRendered` prop to capture the overscan start/stop positions, 
3. Pass those to a `setListPosition` or similar action
4. Have a listener or loader that does a) debounce and b) checks if you're still within the overscan or need to make another api call.
5. Store the results as needed

We will lose the functionality of caching already downloaded parts with this approach, which the infinite loader provides out of the box, but that's not hard to add on top.

And as one more point, I'd move all the static data filtering into a logic or into multiple logics (one per tab?) as well.

**Comment by samwinslow** [comment]
I realized there's a lot of property chaining going on — passing props down again and again to child components. Will move to a structure whereby each component interacts in a flatter way with the data store, via selectors.

This is good, I'm finally understanding the huge benefits of Kea.

**Comment by samwinslow** [comment]
Fuzzy search will require some backend work. Django Postgres weighted queries may be useful, but then again, Clickhouse might be more important to focus on. I will need to find out more about the differences between search implementations on CH vs psql.

Looks promising for Postgres: https://docs.djangoproject.com/en/3.0/ref/contrib/postgres/search/#postgresql-fts-weighting-queries

File: frontend/src/lib/components/PropertyFilters/infiniteListLogic.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            },
+        ],
+        itemsLoading: [
+            false,
+            {
+                setItemsLoading: (_, { itemsLoading }) => itemsLoading,
+            },
+        ],
+    }),
+
```
[the remark this answers is not in the document]
Comment (samwinslow) [reply frontend/src/lib/components/PropertyFilters/infiniteListLogic.ts]: Cleared up the itemsLoading pattern, though I'm not sure if the searchQuery reducer is also problematic?

File: frontend/src/scenes/insights/BreakdownFilter/BreakdownFilter.js (hunk 2 above)
Comment (samwinslow) [reply frontend/src/scenes/insights/BreakdownFilter/BreakdownFilter.js:113]: Resolved — realized this is not the same component as the others, although its name is identical

**Comment by samwinslow** [comment]
Regarding remote search (for event property definitions):

> Fuzzy search will require some backend work. Django Postgres weighted queries may be useful, but then again, Clickhouse might be more important to focus on. I will need to find out more about the differences between search implementations on CH vs psql.

For local search (person properties), I can use Fuse. We use this in some types of filters but not all. What is here is an improvement over `TabbedPropertyFilter` but because of the lack of Fuse support, may be seen as a regression from `UnifiedPropertyFilter` in that regard.

This [Google doc](https://docs.google.com/document/d/1D77-78mEAzn1wdPYGFoC6NvkMm-5dSFt7qCWwJ6_Xdk/edit) contains a list of all instances of property filters and their pros/cons... if this is implemented everywhere Tabbed is currently, it would be an improvement.

**Comment by samwinslow** [comment]
UPDATE: Leaving person properties endpoint as-is (not paginated) per previous discussion. Until it is a standalone table, pagination doesn't make sense because of the added inefficiency on the DB.

**Review by mariusandra — CHANGES_REQUESTED** [review]
I went through the code and added a bunch of inline comments. 

They are all around the same main theme that I've been hammering on for a while, and which you could have predicted and fixed beforehand: put the data in the logic, and don't use `useState`, `useEffect` and other calls to manipulate data. Same goes for e.g. filtering data inside a react component (all that fuse stuff). Just no. Keep data and its manipulations in one place, and views/templates/components in another place.

It's already **a lot** better compared to a few iterations ago, but there is still some funky loopy logic, where a component mounts a logic, then renders, and after asks again the logic to fetch its initial data via useEffect. 

In case there is really no way around some of those calls, that's also fine, but then it should be clearly communicated (e.g. with a ":TRICKY:" comment or whatnot) as to why is it like this.

Then some other comments:

1. I assume only the global filter should be of the new type and the event and breakdown filters not?


2. Searching doesn't work: (I ran `migrate` and also `pip install`)


3. There key errors should all be fixed as well:
image

I'm willing to work on this myself to get it over the line if you'd like me to... or if something is not clear or too complex to refactor. Just let me know if you'd prefer that or want to give it another go. It's however time to get this feature out :).

**Comment by mariusandra** [comment]
Going to close in favour of https://github.com/PostHog/posthog/pull/5098

**Pull request closed without merging**